src/twisted/internet/cfreactor.py

```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
         self._scheduleSimulate()
         return result
 
+    def startRunning(self, installSignalHandlers: bool = True) -> None:
+        """
+        Start running the reactor, then kick off the timer that advances
+        Twisted's clock to keep pace with CFRunLoop's.
+        """
+        super().startRunning(installSignalHandlers)
+        # We must force the simulator to run because in addition to scheduling
+        # timed calls, super().startRunning also fires startup events which may
+        # crash the reactor.
+        self._scheduleSimulate(force=True)
+
     _inCFLoop = False
 
-    def mainLoop(self):
+    def mainLoop(self) -> None:
         """
         Run the runner (C{CFRunLoopRun} or something that calls it), which runs
         the run loop until C{crash()} is called.
         """
-        self._inCFLoop = True
+        # we were crashed during startup.
+        if not self._started:
+
```

Comment:
I'd put this comment inside the `if` since that's the case it's describing.  Also I tried to expand this one and didn't succeed:

```
            # startRunning set _started to True so between then and now,                                                                                                          
            # something crashed us.  There aren't a lot of candidates for                                                                                                         
            # this.  The most well-known is:                                                                                                                                      
            #                                                                                                                                                                     
            #   reactor.callWhenRunning(reactor.crash)                                                                                                                            
            #                                                                                                                                                                     
            # We *don't* eagerly return here because ...                                                                                                                          
            #                                                                                                                                                                     
            #    why?                                                                                                                                                             
            #                                                                                                                                                                     
            # To avoid that, arrange to let the CF main loop run very briefly.                                                                                                    
```

Reply:
OK I think that should do it